go/pkg/cas/upload.go

```diff
@@ -39,6 +65,361 @@ type DigestStat struct {
+//    Rewinding helps locality: there is no delay between reading the file for
+//    the first and the second times.
+//    Only one large file is processed at a time because most GCE disks are
+//    network disks. Reading many large files concurrently appears to saturate
+//    the network and slows down the progress.
+//    See also ClientConfig.LargeFileThreshold.
+func (u *uploader) visitRegularFile(ctx context.Context, absName string, info os.FileInfo) (*repb.FileNode, error) {
+	isLarge := info.Size() >= u.LargeFileThreshold
+
+	// Lock the mutex before acquiring a semaphore to avoid hogging the latter.
+	if isLarge {
+		// Read only one large file at a time.
+		u.muLargeFile.Lock()
+		defer u.muLargeFile.Unlock()
+	}
+
+	if err := u.fsSem.Acquire(ctx, 1); err != nil {
+		return nil, err
+	}
+	defer u.fsSem.Release(1)
+
+	f, err := os.Open(absName)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	ret := &repb.FileNode{
+		Name:         info.Name(),
+		Digest:       &repb.Digest{SizeBytes: info.Size()},
+		IsExecutable: (info.Mode() & 0100) != 0,
+	}
+
+	if info.Size() <= u.SmallFileThreshold {
+		// This file is small enough to buffer it entirely.
+		contents, err := ioutil.ReadAll(f)
+		if err != nil {
+			return nil, err
+		}
+		item := uploadItemFromBlob(absName, contents)
+		ret.Digest = item.Digest
+		return ret, u.scheduleCheck(ctx, item)
+	}
+
+	// It is a medium or large file.
+
+	// Compute the hash.
+	h := digest.HashFn.New()
+	// TODO(nodir): reuse the buffer.
+	buf := make([]byte, u.FileIOSize)
+	if _, err := io.CopyBuffer(h, f, buf); err != nil {
+		return nil, err
+	}
+	ret.Digest.Hash = hex.EncodeToString(h.Sum(nil))
+
+	item := &uploadItem{
+		Title:  absName,
+		Digest: ret.Digest,
+	}
+
+	if isLarge {
+		// Large files are special: locality is important - we want to re-read the
+		// file ASAP.
+		// Also we are not going to use BatchUploads anyway, so we can take
+		// advantage of ByteStream's built-in presence check.
+
+		item.Open = func() (readSeekCloser, error) {
+			_, err := f.Seek(0, io.SeekStart)
+			return f, err
+		}
+		panic("not implemented")
+		// TODO(nodir): implement.
+	}
+
+	// Schedule a check and close the file (in defer).
+	// item.open will reopen the file.
+
+	item.Open = func() (readSeekCloser, error) {
+		return os.Open(absName)
+	}
+	return ret, u.scheduleCheck(ctx, item)
+}
+
+// visitDir reads a directory and its descendants, subject to the
+// predicate. The function blocks until each descendant is visited, but the
+// visitation happens concurrently, using u.eg.
+func (u *uploader) visitDir(ctx context.Context, absName string, predicate FilePredicate) (*repb.DirectoryNode, error) {
+	var mu sync.Mutex
+	dir := &repb.Directory{}
+	var subErr error
+	var wg sync.WaitGroup
+
+	// This sub-function exist to avoid holding the semaphore while waiting for
+	// children.
+	err := func() error {
+		if err := u.fsSem.Acquire(ctx, 1); err != nil {
+			return err
```

Comment:
better to wrap error from standard library?

Reply:
done